File: src/main/java/work/nekow/particledrawing/core/easing/EasingCurve.java
package work.nekow.particledrawing.core.easing;

/**
 * Cubic-bezier easing curve, compatible with CSS easing functions.
 * Defined by four control points where P0=(0,0) and P3=(1,1) are fixed.
 * P1=(x1,y1) and P2=(x2,y2) are user-defined in the [0,1] range.
 *
 * <p>Evaluation uses bisection with Newton-Raphson refinement for O(log N) precision.
 */
@SuppressWarnings("unused")
public final class EasingCurve {

    private static final double EPSILON = 1e-7;
    private static final int MAX_ITERATIONS = 20;

    private final double x1, y1, x2, y2;
    private final double[] sampleCache;

    public EasingCurve(double x1, double y1, double x2, double y2) {
        this.x1 = x1;
        this.y1 = y1;
        this.x2 = x2;
        this.y2 = y2;

        this.sampleCache = new double[11];
        for (int i = 0; i <= 10; i++) {
            sampleCache[i] = sampleCurveX(i / 10.0);
        }
    }

    /**
     * Evaluates the easing curve at progress t in [0, 1].
     * Returns the eased output value in [0, 1].
     */
    public float evaluate(float t) {
        if (t <= 0f) return 0f;
        if (t >= 1f) return 1f;
        return (float) sampleCurveY(solveTForX(t));
    }

    /**
     * Returns the control point values, suitable for serialization.
     */
    public double x1() { return x1; }
    public double y1() { return y1; }
    public double x2() { return x2; }
    public double y2() { return y2; }

    private double sampleCurveX(double t) {
        return ((1 - t) * (1 - t) * (1 - t) * 0)
             + 3 * (1 - t) * (1 - t) * t * x1
             + 3 * (1 - t) * t * t * x2
             + t * t * t * 1;
    }

    private double sampleCurveY(double t) {
        return ((1 - t) * (1 - t) * (1 - t) * 0)
             + 3 * (1 - t) * (1 - t) * t * y1
             + 3 * (1 - t) * t * t * y2
             + t * t * t * 1;
    }

    private double sampleCurveDerivativeX(double t) {
        return 3 * (1 - t) * (1 - t) * (x1 - 0)
             + 6 * (1 - t) * t * (x2 - x1)
             + 3 * t * t * (1 - x2);
    }

    private double solveTForX(double x) {
        double t = x;
        for (int i = 0; i < 8; i++) {
            double curX = sampleCurveX(t) - x;
            if (Math.abs(curX) < EPSILON) {
                return t;
            }
            double d = sampleCurveDerivativeX(t);
            if (Math.abs(d) < 1e-6) {
                break;
            }
            t = t - curX / d;
        }

        double t0 = 0;
        double t1 = 1;
        t = x;

        if (t < t0) return t0;
        if (t > t1) return t1;

        for (int i = 0; i < MAX_ITERATIONS; i++) {
            double curX = sampleCurveX(t) - x;
            if (Math.abs(curX) < EPSILON) {
                return t;
            }
            if (curX > 0) {
                t1 = t;
            } else {
                t0 = t;
            }
            t = (t0 + t1) / 2.0;
        }

        return t;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof EasingCurve c)) return false;
        return Double.compare(c.x1, x1) == 0
            && Double.compare(c.y1, y1) == 0
            && Double.compare(c.x2, x2) == 0
            && Double.compare(c.y2, y2) == 0;
    }

    @Override
    public int hashCode() {
        return java.util.Objects.hash(x1, y1, x2, y2);
    }

    @Override
    public String toString() {
        return String.format("cubic-bezier(%.3f, %.3f, %.3f, %.3f)", x1, y1, x2, y2);
    }

    public static EasingCurve fromCss(String css) {
        String inner = css.replace("cubic-bezier(", "").replace(")", "").trim();
        String[] parts = inner.split(",");
        if (parts.length != 4) {
            throw new IllegalArgumentException("Invalid CSS cubic-bezier: " + css);
        }
        return new EasingCurve(
            Double.parseDouble(parts[0].trim()),
            Double.parseDouble(parts[1].trim()),
            Double.parseDouble(parts[2].trim()),
            Double.parseDouble(parts[3].trim())
        );
    }
}

File: src/main/kotlin/work/nekow/particledrawing/core/easing/EasingCurve.kt
package work.nekow.particledrawing.core.easing

import java.util.Objects
import kotlin.math.abs

/**
 * 三次贝塞尔缓动曲线，由四个控制点参数定义。
 * 用于在粒子动画中计算平滑的过渡效果。
 *
 * @param x1 第一控制点的 X 坐标
 * @param y1 第一控制点的 Y 坐标
 * @param x2 第二控制点的 X 坐标
 * @param y2 第二控制点的 Y 坐标
 */
@Suppress("unused")
class EasingCurve(
    val x1: Double,
    val y1: Double,
    val x2: Double,
    val y2: Double
) {

    // 线性曲线（默认缓动）：直接返回 t，避免贝塞尔反解迭代
    private val linear = x1 == 0.0 && y1 == 0.0 && x2 == 1.0 && y2 == 1.0

    /**
     * 在给定进度下计算缓动值。
     * @param t 动画进度，范围 [0, 1]
     * @return 缓动后的值，范围 [0, 1]
     */
    fun evaluate(t: Float): Float {
        if (t <= 0f) return 0f
        if (t >= 1f) return 1f
        if (linear) return t
        return sampleCurveY(solveTForX(t.toDouble())).toFloat()
    }

    private fun sampleCurveX(t: Double): Double {
        val u = 1 - t
        return 3 * u * u * t * x1 + 3 * u * t * t * x2 + t * t * t
    }

    private fun sampleCurveY(t: Double): Double {
        val u = 1 - t
        return 3 * u * u * t * y1 + 3 * u * t * t * y2 + t * t * t
    }

    private fun sampleCurveDerivativeX(t: Double): Double {
        val u = 1 - t
        return 3 * u * u * x1 + 6 * u * t * (x2 - x1) + 3 * t * t * (1 - x2)
    }

    private fun solveTForX(x: Double): Double {
        var t = x
        for (i in 0 until 8) {
            val curX = sampleCurveX(t) - x
            if (abs(curX) < EPSILON) {
                return t
            }
            val d = sampleCurveDerivativeX(t)
            if (abs(d) < 1e-6) {
                break
            }
            t = (t - curX / d).coerceIn(0.0, 1.0)
        }

        var t0 = 0.0
        var t1 = 1.0
        for (i in 0 until MAX_ITERATIONS) {
            t = (t0 + t1) / 2.0
            if (sampleCurveX(t) - x < 0) {
                t0 = t
            } else {
                t1 = t
            }
        }
        return t
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is EasingCurve) return false
        return x1.compareTo(other.x1) == 0
            && y1.compareTo(other.y1) == 0
            && x2.compareTo(other.x2) == 0
            && y2.compareTo(other.y2) == 0
    }

    override fun hashCode(): Int {
        return Objects.hash(x1, y1, x2, y2)
    }

    override fun toString(): String {
        return String.format("cubic-bezier(%.3f, %.3f, %.3f, %.3f)", x1, y1, x2, y2)
    }

    companion object {
        private const val EPSILON = 1e-7
        private const val MAX_ITERATIONS = 20

        /**
         * 从 CSS 格式的字符串解析缓动曲线。
         * @param css CSS cubic-bezier 格式字符串
         * @return 解析后的 EasingCurve 实例
         */
        @JvmStatic
        fun fromCss(css: String): EasingCurve {
            val inner = css.replace("cubic-bezier(", "").replace(")", "").trim()
            val parts = inner.split(",")
            require(parts.size == 4) { "Invalid CSS cubic-bezier: $css" }
            return EasingCurve(
                parts[0].trim().toDouble(),
                parts[1].trim().toDouble(),
                parts[2].trim().toDouble(),
                parts[3].trim().toDouble()
            )
        }
    }
}
